### PR TITLE
[extractor/vocaroo] Add Vocaroo extractor

### PR DESCRIPTION
**Pull request by qbnu**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

(Description from https://github.com/ytdl-org/youtube-dl/pull/29369)

Adds support for [Vocaroo](https://www.vocaroo.com/), "the premier voice recording service." The site lacks any sort of metadata for its uploads and all come down as MP3s, so much of this extractor is hardcoded, making it quite fast. While Vocaroo does allow you to easily download individual files from their site, youtube-dl support will allow users to extract Vocaroo embeds with the generic extractor and download many Vocaroo files at once.

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

### CODE REVIEW

File: yt_dlp/extractor/vocaroo.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    ]
+
+    def _real_extract(self, url):
+        audio_id = self._match_id(url)
+
+        return {
+            'id': audio_id,
+            'title': f'Vocaroo {audio_id}',
+            'url': f'https://media1.vocaroo.com/mp3/{audio_id}',
+            'ext': 'mp3',
```
Comment (pukkandan) [inline yt_dlp/extractor/vocaroo.py:33]:
```suggestion
            'ext': 'mp3',
            'vcodec': 'none',
```

File: yt_dlp/extractor/vocaroo.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+            'only_matching': True,
+        },
+    ]
+
+    def _real_extract(self, url):
+        audio_id = self._match_id(url)
+
+        return {
+            'id': audio_id,
+            'title': f'Vocaroo {audio_id}',
```
Comment (pukkandan) [inline yt_dlp/extractor/vocaroo.py:31]:
```suggestion
            'title': '',
```

File: yt_dlp/extractor/vocaroo.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
+from .common import InfoExtractor
+
+
+class VocarooIE(InfoExtractor):
+    _VALID_URL = r'https?://(?:www\.)?(?:vocaroo\.com|voca\.ro)/(?:embed/)?(?P<id>[a-zA-Z0-9]{11,12})'
```
Comment (pukkandan) [inline yt_dlp/extractor/vocaroo.py:5]:
```suggestion
    _VALID_URL = r'https?://(?:www\.)?(?:vocaroo\.com|voca\.ro)/(?:embed/)?(?P<id>\w+)'
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

**Comment by pukkandan** [comment]
> youtube-dl support will allow users to extract Vocaroo embeds with the generic extractor and download many Vocaroo files at once.

The current extractor doesn't seem to support embeds. Provide example

**Comment by qbnu** [comment]
There is an `x-bz-upload-timestamp` header in the response which has the UNIX timestamp of when each file was uploaded. Should this be used as the value for `timestamp`, and is there a way to do it without sending an extra request?

**Comment by dirkf** [comment]
You can can make a `HEAD` request in the extractor (or `GET` if the site doesn't support it) and get the header value out of the response. Any other solution would be overkill.

Actually, sending the request might be a good thing. The extractor currently just guesses that an audio file exists with the given ID, whereas making at least one request will reveal whether the ID is valid (or perhaps that the extractor's guess is no longer valid).

In my test I found that a `media1....` URL redirected to `media....`.

**Comment by qbnu** [comment]
I made the logic for the media subdomain match the logic of the site
```js
function (e) {
    if (e.length) {
        if (11 == e.length)
            return ControlConfig.mediaMp3FileUrl;
        if (12 == e.length && '1' == e[0])
            return ControlConfig.mediaMp3FileUrl1;
        if (10 == e.length)
            return ControlConfig.mediaMp3FileUrl1
    }
    return ControlConfig.mediaMp3FileUrl
}
```
> In my test I found that a `media1....` URL redirected to `media....`.

Can you give an example? I assumed they were mirrors.

File: yt_dlp/extractor/vocaroo.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        audio_id = self._match_id(url)
+        if len(audio_id) == 10 or (len(audio_id) == 12 and audio_id[0] == '1'):
+            media_subdomain = 'media1'
+        else:
+            media_subdomain = 'media'
+
+        url = f'https://{media_subdomain}.vocaroo.com/mp3/{audio_id}'
+        http_headers = {'Referer': 'https://vocaroo.com/'}
+        req = HEADRequest(url, headers=http_headers)
+        resp = self._request_webpage(req, audio_id)
```
Comment (pukkandan) [inline yt_dlp/extractor/vocaroo.py:57]:
```suggestion
        resp = self._request_webpage(HEADRequest(url), audio_id, headers=http_headers)
```

File: yt_dlp/extractor/vocaroo.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        timestamp = float_or_none(resp.getheader('x-bz-upload-timestamp'), scale=1000)
+
+        return {
+            'id': audio_id,
+            'title': '',
+            'url': url,
+            'ext': 'mp3',
+            'timestamp': timestamp,
+            'vcodec': 'none',
+            'http_headers': http_headers,
```
Comment (pukkandan) [inline yt_dlp/extractor/vocaroo.py:67]:
```suggestion
        return {
            'id': audio_id,
            'title': '',
            'url': url,
            'ext': 'mp3',
            'timestamp': float_or_none(resp.getheader('x-bz-upload-timestamp'), scale=1000),
            'vcodec': 'none',
            'http_headers': http_headers,
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

**Comment by SuperSonicHub1** [comment]
Hey! Looks like some more of my code made it into yt-dlp. I have no issue with my code being used without permission—it is public domain after all—but being mentioned would have been nice; would've been happy to help and would have closed the PR in youtube-dl. That's all; sorry for waking a sleeping thread.

**Comment by pukkandan** [comment]
Darn it. I do mention usually, but I didn't notice for this since OP said:

> * [x]  I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)

Too late to do anything about commit message now, but I'll add you to changelog

**Comment by pukkandan** [comment]
@SuperSonicHub1 You are credited in commit message and changelog

**Comment by SuperSonicHub1** [comment]
Thanks, @pukkandan!

**Pull request merged**